Zend/zend_partial.c

```diff
@@ -0,0 +1,920 @@
+	zend_object *object = NULL;
+	zval *params;
+	uint32_t num_params, num_all_params;
+	HashTable *named_args;
+	zend_fcall_info fci = empty_fcall_info;
+	zend_fcall_info_cache fcc = empty_fcall_info_cache;
+
+	ZEND_PARSE_PARAMETERS_START(0, -1)
+		Z_PARAM_VARIADIC_WITH_NAMED(params, num_params, named_args)
+	ZEND_PARSE_PARAMETERS_END();
+
+	num_all_params = num_params + (named_args ? named_args->nNumUsed : 0);
+
+	if (num_all_params < partial->prototype.common.required_num_args) {
+		zend_string *symbol = zend_partial_symbol_name(execute_data, &partial->prototype);
+		zend_partial_prototype_underflow(
+			&partial->prototype, symbol, num_all_params,
+			partial->prototype.common.required_num_args, 
+			ZEND_PARTIAL_CALL_FLAG(partial, ZEND_APPLY_VARIADIC));
+		zend_string_release(symbol);
+		if (ZEND_PARTIAL_IS_CALL_TRAMPOLINE(&partial->prototype)) {
+			EG(trampoline).common.function_name = NULL;
+		}
+		RETURN_THROWS();
+	} else if (num_all_params > partial->prototype.common.num_args && 
+			  !ZEND_PARTIAL_CALL_FLAG(partial, ZEND_APPLY_VARIADIC)) {
+		zend_string *symbol = zend_partial_symbol_name(execute_data, &partial->prototype);
+		zend_partial_prototype_overflow(
+			&partial->prototype, symbol, num_all_params,
+			partial->prototype.common.num_args);
+		zend_string_release(symbol);
+		if (ZEND_PARTIAL_IS_CALL_TRAMPOLINE(&partial->prototype)) {
+			EG(trampoline).common.function_name = NULL;
+		}
+		RETURN_THROWS();
+	}
+
+	fci.size = sizeof(zend_fcall_info);
+
+	if (UNEXPECTED(ZEND_PARTIAL_CALL_FLAG(partial, ZEND_APPLY_FACTORY))) {
+		zend_class_entry *type = Z_CE(partial->This);
+		zval instance;
+
+		if (type->create_object) {
+			ZVAL_OBJ(&instance, type->create_object(type));
+		} else {
+			ZVAL_OBJ(&instance, zend_objects_new(type));
+
+			object_properties_init(Z_OBJ(instance), type);
+		}
+
+		object = Z_OBJ(instance);
+
+		if (ZEND_PARTIAL_CALL_FLAG(partial, ZEND_APPLY_PASS)) {
+			/* nothing to pass any arguments too, so return immediately */
+			RETURN_OBJ(object);
+		}
+
+		GC_ADD_FLAGS(object, IS_OBJ_DESTRUCTOR_CALLED);
+	} else {
+		if (Z_TYPE(partial->This) == IS_OBJECT) {
+			object = Z_OBJ(partial->This);
+		} else if (Z_TYPE(partial->This) == IS_UNDEF && Z_CE(partial->This)) {
+			fcc.called_scope = Z_CE(partial->This);
+		}
+	}
+
+	fci.retval = return_value;
+	fcc.function_handler = &partial->func;
+
+	if (object) {
+		fci.object = object;
+		fcc.object = object;
+	}
+
+	if (partial->named) {
+		if (!named_args) {
+			named_args = partial->named;
+			GC_ADDREF(named_args);
+		} else {
+			HashTable *nested = zend_array_dup(partial->named);
+			zend_hash_merge(nested, named_args, zval_copy_ctor, true);
+			named_args = nested;
+		}
+		fci.named_params = named_args;
+	} else {
+		fci.named_params = named_args;
+	}
+
+	fci.params = ecalloc(sizeof(zval), partial->argc + num_params);
+	fci.param_count = zend_partial_apply(
+		partial->argv, partial->argv + partial->argc,
+		params, params + num_params, fci.params, true);
```

Comment:
`params + num_params` causes ubsan violation when params == NULL.

Reply:
Fixed in a89166187bd2b536094f43945da5ca1ae357a3cb